README.md
# homebridge-away-mode

[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)
[![mit license](https://badgen.net/badge/license/MIT/red)](https://github.com/isalkind/homebridge-away-mode/blob/master/LICENSE)
[![npm](https://badgen.net/npm/dt/homebridge-away-mode)](https://www.npmjs.com/package/homebridge-away-mode)

[Homebridge](https://homebridge.io) plugin that provides triggers to turn on and turn off lights to simulate occupancy. We provide the triggers, you provide the lights.

A simulated switch is created that controls whether "away mode" is active. When the switch is on, away mode is active. When the switch is off, away mode is inactive. A set of simulated sensors detect "activity". When activity (motion) is detected, turn the light on. When activity (motion) is not detected, turn the light off. The behavior of each sensor is random: a sensor is off for a period of time, turns on for a period of time, then repeats. When the switch is turned on, the sensors are activated to start their off/on behavior. When the switch is turned off, the sensors are deactivated and turned off.

## Homebridge Restart

The plugin attempts to restore itself to the its previous state if a restart occurs and the "away mode" switch is on:
1. The "away mode" switch is turned on.
1. For any sensor that was previously on:
    1. If the sensor is still allowed to be on, it is turned on and the normal on/off sequence is resumed, starting from on.
    1. If the sensor is not allowed to turn on, it is not turned on and the normal on/off sequence is resumed, starting from off.
1. Any sensor that was previously off will begin its on/off sequence, starting from off.

_**Caveat** - Because the plugin does not track how long a sensor had been on (or off) when a restart occurred, it is possible that the sensor will be on (or off) longer than the settings would normally allow. This will only effect the first sequence following the restart._

# Installation

npm -g install homebridge-away-mode

# Configuration

It is highly recommended to use the [homebridge-config-ui-x](https://www.npmjs.com/package/homebridge-config-ui-x) plugin for ease of configuration. The `homebridge-away-mode` plugin's behavior is driven almost entirely by the values you specify in the configuration. Depending on how many sensors you configure and how complex you make each sensor's behavior, the configuration can become quite extensive.

### Simple configuration example
config.json:
```
"accessories": [
    {
        "accessory": "AwayMode",
        "name": "Away Mode",
        "sensors": [
            {
                "name": "Trigger1"
            },
            {
                "name": "Trigger2"
            }
        ],
        "minOffTime": 300,
        "maxOffTime": 1800,
        "minOnTime": 1800,
        "maxOnTime": 3600
    }
]
```
The above example creates a switch called 'Away Mode' and two sensors: 'Trigger 1' and 'Trigger 2'. A sensor will remain off for 300 to 1800 seconds (5 minutes to 1/2 hour). A sensor will remain on for 1800 to 3600 seconds (1/2 hour to 1 hour). Sensors will only be active when the 'Away Mode' switch is turned on.

In this example, you use your automation software to turn the 'Away Mode' switch on when you want the sensors to be active (turn on & off), and off when you don't want them to be active.
##### HomeKit
In Home (or any of the HomeKit compatible apps), you will see three new devices exposed:
* 1 switch named 'Away Mode'
* 2 motion sensors named 'Trigger 1' and 'Trigger 2'

Automate the switch to turn on and off at specific times. For example:
* 15 minutes before sunset (daily), turn the switch on
* At 10:00pm (daily), turn the switch off

Automate the motions sensors. For Example:
* When motion is detected ('Trigger 1'), turn 'Lamp 1' on
* When no motion is detected ('Trigger 1'), turn 'Lamp 1' off

This setup makes sense when you know you will be away from your house for an extended period of time. You enable the switch automation before you leave and disable it when you return.

### Advanced configuration example
config.json:
```
"accessories": [
    {
        "accessory": "AwayMode",
        "name": "Away Mode",
        "sensors": [
            {
                "name": "Trigger1"
            },
            {
                "name": "Trigger2"
            }
        ],
        "minOffTime": 300,
        "maxOffTime": 1800,
        "minOnTime": 1800,
        "maxOnTime": 3600,
        "activeTimes": [
            {
                "start": "sunset",
                "end": "22:00"
            }
        ],
        "location": {
            "lat": 40.689510,
            "long": -74.044500
        },
        "offset": {
            "sunrise": 0,
            "sunset": -15
        }
    }
]
```
The above example creates a switch called 'Away Mode' and two sensors: 'Trigger 1' and 'Trigger 2'. A sensor will remain off for 300 to 1800 seconds (5 minutes to 1/2 hour). A sensor will remain on for 1800 to 3600 seconds (1/2 hour to 1 hour). Sensors will only be active when the 'Away Mode' switch is turned on. Sensors will only turn on from 15 minutes before sunset to 10:00pm. The location information is used to compute the values for sunrise/sunset (as needed).

In this example, you might use your automation software to detect when your residence is not occupied and turn on the 'Away Mode' and turn if off when the residence is occupied. However, you only want the sensors turning on and off during a specific period of the day.

**Note**: This can *mostly* be accomplished using HomeKit and the simple configuration. However, HomeKit does not allow you to mix specific times (eg., 10:00pm) with sunrise/sunset times.

##### HomeKit
In Home (or any of the HomeKit compatible apps), you will see three new devices exposed:
* 1 switch named 'Away Mode'
* 2 motion sensors named 'Trigger 1' and 'Trigger 2'

Automate the switch to turn on and off based on specific conditions. For example:
* When the last person leaves home, turn the switch on
* When the first person arrives home, turn the switch off

Automate the motions sensors. For Example:
* When motion is detected ('Trigger 1'), turn 'Lamp 1' on
* When no motion is detected ('Trigger 1'), turn 'Lamp 1' off

This setup makes sense when you want to make sure the automation always runs when you are not home, but only want the sensors to trigger during specific times of the day.

### Parameters

| Param         | Description   | Default  |
| ------------- | ------------- | -------- |
| accessory | **Must be set to "AwayMode"**
| name | The name of the switch | "Away Mode" |
| sensorNames | Array of names for each sensor to be created<br><br>**DEPRECATED** - See 'sensors' parameter. If the 'sensors' parameter is specified, this will be ignored. | ["Trigger 1"] |
| sensors | Array of per-sensor information.<br><br>Each sensor object **MUST** contain 'name' and **MAY** contain 'minOffTime', 'maxOffTime', 'minOnTime', 'maxOnTime', 'activeTimesForSensor'. These *time* parameters override the globally defined *time* parameters if specified.<br><br>Example 1: [{"name": "Trigger1"}]<br>Set sensor name, use the global time parameters.<br><br>Example 2: [{"name": "Trigger1", "minOffTime": 27}]<br>Set sensor name, override the 'minOffTime' global time parameter.<br><br>Example 3: [{"name": "Trigger1", "activeTimesForSensor":[{"start": "sunset", "end": "23:00"}]}]<br>Set sensor name, override the 'activeTimes' global time parameter.<br><br>See global time parameters: 'minOffTime', 'minOffTime', 'minOnTime', 'maxOffTime', 'activeTimes'<br><br>**Note:** The 'activeTimesForSensor' parameter maps to the global 'activeTimes' parameter.<br><br>**Note:** If the 'sensors' parameter is specified, the 'sensorNames' parameter will be ignored. | [{"name": "Trigger1"}] |
| minOffTime | Minimum off time (secs) | 300 |
| maxOffTime | Maximum off time (secs) | 1800 |
| minOnTime | Minimum on time (secs) | 1800 |
| maxOnTime | Maximum on time (secs) | 3600 |
| startTime | Time at which triggers should start to fire<br>("hh:mm"\|"sunrise"\|"sunset")<br><br>**DEPRECATED** - See 'activeTimes' parameter. If the 'activeTimes' parameter is specified, this will be ignored. | "00:00" |
| endTime | Time at which triggers should stop firing<br>("hh:mm"\|"sunrise"\|"sunset")<br><br>**DEPRECATED** - See 'activeTimes' parameter. If the 'activeTimes' parameter is specified, this will be ignored. | "23:59" |
| activeTimes | Array of start/end times for periods when triggers should fire.<br>Set start/end times as: ("hh:mm"\|"sunrise"\|"sunset")<br><br>You may also include the optional "absolute" parameter as: (true\|false). When set to true, sensors will turn off immediately when the end time is reached (if one is on).<br><br>Example 1: [{"start": "sunset", "end": "22:00"}]<br><br>Example 2: [{"start": "sunset", "end": "22:00", "absolute": true}]| [{"start": "00:00", "end": "23:59"}] |
| location | Lat/long location to compute sunrise/sunset from. Use in conjunction with "startTime"\|"endTime" when they are set to "sunrise"\|"sunset".<br>({"lat": x, "long": y}) <br><br>Find your location: [Google Maps location finder](https://google-developers.appspot.com/maps/documentation/utils/geocoder/) | {"lat": 0, "long": 0} |
| offset | Offset information for sunrise/sunset. Offset is in minutes. May be negative (before) or positive (after). Use in conjunction with "startTime"\|"endTime" when they are set to "sunrise"\|"sunset".<br>({"sunrise": mins, "sunset": mins}) | {"sunrise": 0, "sunset": 0} |

Parameters in the table above are optional, except 'accessory' which must be set to 'AwayMode', and need not be specified if you are happy with the default value.

Time examples:

| startTime | endTime | Offset |Description |
| --------- | ------- | ------ | ----------- |
| "08:00" | "20:00" | NA | Span the hours from 8am to 8pm |
| "20:00" | "08:00" | NA | Span the hours from 8pm to 8am |
| "sunset" | "22:00" | {"sunset":-15} | Span the hours from 15 minutes before sunset to 10pm. Sunset is computed based on the location you provide. |
| "23:00" | "sunrise" | {"sunrise":30} | Span the hours from 11pm to 30 minutes after sunrise. Sunrise is computed based on the location you provide. |
| "sunrise" | "sunset" | {"sunrise":-15, "sunset":15} | Span the hours from 15 minutes before sunrise to 15 minutes after sunset. Sunrise and sunset are computed based on the location you provide. |


CHANGELOG.MD

# Change Log

All notable changes to this project will be documented in this file.

## 1.5.2 (2020-12-08)

### Notable Changes

* Restore sensors to their previous state after restart if homebridge was restarted while the 'Away Mode' switch was on.

## 1.5.1 (2020-11-28)

### Notable Changes

* Maintain switch state across restarts. If the homebridge is restarted while the 'Away Mode' switch is on, activities will be resumed after the restart. ([#8](https://github.com/isalkind/homebridge-away-mode/issues/8))

## 1.5.0 (2020-05-11)

### Notable Changes

* Allow specification of active times on a per-sensor basis. See the new 'activeTimesForSensor' configuration parameter that is part of the 'sensors' configuration. Values specified by 'activeTimesForSensor' will override the global values in the 'activeTimes' configuration parameter.

* The active times configuration now allows an optional parameter to be added to each active period to force any sensors to turn off when the end of the active period is reached. The default behavior allows the sensor to remain on until it naturally turns off. See the 'activeTimes' configuration parameter and optional 'absolute' element. This optional configuration is also available on the per-sensor override parameter 'activeTimesForSensor'. ([#7](https://github.com/isalkind/homebridge-away-mode/issues/7))

## 1.4.2 (2020-01-12)

### Notable Changes

* Updated CHANGELOG. Forgot to do it on 1.4.1.

## 1.4.1 (2020-01-12)

### Bug Fixes

* Shim padStart for String on early versions of node. ([#5](https://github.com/isalkind/homebridge-away-mode/issues/5))

## 1.4.0 (2020-01-06)

### Notable Changes

* Adopt the config.schema.json used by [Homebridge Config UI X](https://github.com/oznu/homebridge-config-ui-x) to perform web-based configuration for this plugin.

### Other Changes

* Add **Change Log**

## 1.3.0 (2019-12-15)

### Notable Changes

* Configure the min/max on and off times on a per-sensor basis. See the new 'sensors' configuration parameter. ([#4](https://github.com/isalkind/homebridge-away-mode/issues/4))

## 1.2.0 (2019-12-02)

### Notable Changes

* Specify multiple time ranges for triggers to activate. See 'activeTimes' configuration parameter. ([#2](https://github.com/isalkind/homebridge-away-mode/issues/2))

## 1.1.0 (2019-02-23)

### Notable Changes

* Specify a time range for triggers to be active. See 'startTime' and 'endTime' configuration parameters.
* Specify location to compute 'sunrise' and 'sunset' values from. See 'location' configuration parameter.
* Specify offset information for 'sunrise' and 'sunset'. See 'offset' configuration parameter.

### Other Changes

* Updated npm dependencies

## 1.0.0 (2018-01-30)

### Initial Release


index.js
////////////////////////////////////////////////////////////////////////////////
//
// AwayMode - Homebridge plugin that provides triggers to turn on and turn off
//            lights to simulate occupancy. We provide the triggers, you provide
//            the lights.
//
// A simulated switch is created that controls whether "away mode" is active.
// When the switch is on, away mode is active. When the switch is off, away mode
// is inactive. A set of simulated sensors detect "activity". When activity
// (motion) is detected, turn the light on. When activity (motion) is not
// detected, turn the light off. The behavior of each sensor is random: a sensor
// is off for a period of time, turns on for a period of time, then repeats.
// When the switch is turned on, the sensors are activiated to start their
// off/on behavior. When the switch is turned off, the sensors are deactived
// and turned off.
//
// Config - (defaults in parenthesis)
//
// name - The name of the switch. ("Away Mode")
// sensorNames - Array of names for each sensor to be created. (["Trigger 1"])
// sensors - Array of per-sensor information.
// minOffTime - Minimum off time (seconds). (300)
// maxOffTime - Maximum off time (seconds). (1800)
// minOnTime - Minimum on time (seconds). (1800)
// maxOnTime - Maximum off time (seconds). (3600)
// startTime - Time at which triggers should start to fire (hh:mm|sunrise|sunset). (00:00)
// endTime - Time at which triggers should stop firing (hh:mm|sunrise|sunset). (23:59)
// activeTimes - Array of start/end times (see startTime/endTime)
// location - lat/long location to compute sunrise/sunset from ({lat: x, long: y}). ({lat: 0, long: 0})
// offset - Offset information for sunrise/sunset ({sunrise: mins, sunset: mins}). ({sunrise: 0, sunset: 0})
//
// User location can be found with:
//     https://google-developers.appspot.com/maps/documentation/utils/geocoder/
//
////////////////////////////////////////////////////////////////////////////////

const storage = require('node-persist');
const TimeFormat = require('hh-mm-ss');
const SunCalc = require('suncalc');
var padStart = require('string.prototype.padstart');

var Homebridge, Service, Characteristic;

module.exports = function(homebridge) {
    Homebridge = homebridge;
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;

    homebridge.registerAccessory("homebridge-away-mode", "AwayMode", AwayMode);
}

class AwayMode {

    //
    // Constructor. Create control switch and sensors.
    //
    constructor(log, config) {
        // Shim padStart if it is unavailable (early versions of node)
        if (!String.prototype.padStart) {
            padStart.shim();
        }

        this.log = log;
        this.name = config["name"] || "Away Mode";
        this.sensorNames = config["sensorNames"] || ["Trigger 1"];
        this.sensors = config["sensors"] || [];

        // Time in seconds
        this.minOffTime = config["minOffTime"] || 300;  // 5 min (secs)
        this.maxOffTime = config["maxOffTime"] || 1800; // 30 min (secs)
        this.minOnTime = config["minOnTime"] || 1800;   // 30 min (secs)
        this.maxOnTime = config["maxOnTime"] || 3600;   // 60 min (secs)

        this.startTime = config["startTime"] || "00:00"; // hh:mm|sunrise|sunset
        this.endTime = config["endTime"] || "23:59";     // hh:mm|sunrise|sunset
        this.activeTimes = config["activeTimes"] || [{start:this.startTime, end:this.endTime}];

        // https://google-developers.appspot.com/maps/documentation/utils/geocoder/
        this.location = config["location"] || {lat:0, long:0};
        this.offset = config["offset"] || {sunrise:0, sunset:0}; // 0 min (mins)

        // Multiplier to get to timer values (in milliseconds)
        this.multiplier = 1000;

        this.isSwitchOn = false;

        // Create switch to turn on/off away mode
        this.serviceSwitch = new Service.Switch(this.name);
        this.log("Switch: " + this.name);

        this.serviceSwitch
        .getCharacteristic(Characteristic.On)
        .on('get', function(callback) {
            callback(null, this.isSwitchOn);
        }.bind(this))
        .on('set', this.setOn.bind(this));

        // Check for sensors. If they don't exist, we
        // will create them here. For backwards compatibility.
        if (this.sensors.length === 0) {
            for (let x = 0; x < this.sensorNames.length; x++) {
                this.sensors[x] = {
                    name: this.sensorNames[x]
                }
            }
        }

        // Create motion sensors that will fire randomly
        this.serviceMotions = [];
        this.serviceStates = [];

        for (let x = 1; x <= this.sensors.length; x++) {
            let sensor = this.sensors[x-1];
            let serviceMotion = new Service.MotionSensor(sensor.name, x);
            this.log("MotionSensor: " + sensor.name);

            let serviceState = { "motionDetected": false };

            serviceMotion.getCharacteristic(Characteristic.MotionDetected)
            .on('get', function(callback) {
                callback(null, serviceState.motionDetected);
            }.bind(this));

            this.serviceMotions.push(serviceMotion);
            this.serviceStates.push(serviceState);

            // populate sensor defaults if necessary
            sensor.minOffTime = sensor.minOffTime || this.minOffTime;
            sensor.maxOffTime = sensor.maxOffTime || this.maxOffTime;
            sensor.minOnTime = sensor.minOnTime || this.minOnTime;
            sensor.maxOnTime = sensor.maxOnTime || this.maxOnTime;

            // populate active times if necessary
            sensor.activeTimesForSensor = sensor.activeTimesForSensor || this.activeTimes;
            sensor.activeSeconds = new Array(sensor.activeTimesForSensor.length);

            // This call computes side-effects - activeSeconds
            this.computeStartEndTimesForSensor(x-1);
        }

        this.log(`Sensors: ${JSON.stringify(this.sensors)}`);

        // Restore previous state as necessary
        setTimeout(this.restore.bind(this));
    }

    async restore() {
        // initialize local storage
        await storage.init({dir: Homebridge.user.persistPath(), forgiveParseErrors: true});

        // retrieve previous state - main switch
        let active = await storage.getItem('active');

        // main switch previously on
        if (active === 'true') {
            this.log('Restore switch state to on');
            this.serviceSwitch.updateCharacteristic(Characteristic.On, true);
            // set the restore flag on sensors so we check them on first activation
            for (let i=0; i<this.sensors.length; i++) {
                this.sensors[i].restore = true;
            }
            this.setOn(true, undefined);
        }
    }

    //
    // Convert seconds to "hh:mm:ss" for display
    //
    secondsToHourMinSec(timeInSeconds) {
        let r = 0;

        // hours
        let hours = Math.floor(timeInSeconds / 3600);

        // minutes
        r = timeInSeconds % 3600
        let mins = Math.floor(r / 60);

        // seconds
        r = r % 60;

        return '' + hours.toString().padStart(2, '0') + ':' +
               mins.toString().padStart(2, '0') + ':' +
               r.toString().padStart(2, '0');
    }

    //
    // Compute the seconds from midnight for the given
    // time specification. Time is either 'hh:mm', 'sunrise', or 'sunset'.
    // 'sc' contains sunrise/sunset info. 'offset' only applies to
    // sunrise/sunset calculations.
    //
    computeSecondsFromMidnight(sc, time, offset) {
      let seconds = 0;
      let dynamic = false;

      if (time === 'sunrise' || time === 'sunset') {
          dynamic = true;
          let t = 0; // base time - sunrise or sunset
          let o = 0; // offset from t
          if (time === 'sunrise') {
              t = sc.sunrise;
              o = offset.sunrise * 60; // to seconds
          } else {
              t = sc.sunset;
              o = offset.sunset * 60; // to seconds
          }
          seconds = (t.getHours() * 3600) + (t.getMinutes() * 60) + t.getSeconds() + o;
      } else {
          seconds = TimeFormat.toS(time, 'hh:mm');
      }

      return { seconds: seconds, dynamic: dynamic };
    }

    //
    // Compute start and end times. If dynamic values present (sunrise | sunset),
    // this method will be called at midnight every day.
    //
    // *** Side effects ***
    // Sets the values for sensor.activeSeconds
    //
    computeStartEndTimesForSensor(id) {
        const sensor = this.sensors[id];
        let dynamic = false;

        let times = SunCalc.getTimes(new Date(), this.location.lat, this.location.long);

        // Examine each of the defined ranges
        for (let i=0; i<sensor.activeTimesForSensor.length; i++) {
            let startInfo = this.computeSecondsFromMidnight(times, sensor.activeTimesForSensor[i].start, this.offset);
            let endInfo = this.computeSecondsFromMidnight(times, sensor.activeTimesForSensor[i].end, this.offset);
            dynamic = dynamic || startInfo.dynamic || endInfo.dynamic;

            sensor.activeSeconds[i] = {start: startInfo.seconds, end: endInfo.seconds};

            this.log("[" + sensor.name + "]" + "[" + i + "] Start: " + this.secondsToHourMinSec(sensor.activeSeconds[i].start) +
                     " End: " + this.secondsToHourMinSec(sensor.activeSeconds[i].end));
        }

        // Set a timer to expire at midnight so we can recalculate the
        // values of sunrise & sunset (if needed)
        if (dynamic) {
            let today = new Date();
            let tommorow = new Date(today.getFullYear(),today.getMonth(),today.getDate()+1);
            let timeToMidnight = (tommorow-today)+5000; // +5sec to push it past
            this.log("Recompute in: " + timeToMidnight/1000);
            let timer = setTimeout(function() {
                this.computeStartEndTimesForSensor(id);
            }.bind(this), timeToMidnight);
        }
    }

    //
    // Representation of current time in seconds.
    //
    currentSeconds() {
        let now = new Date();
        return (now.getHours() * 3600) + (now.getMinutes() * 60) + now.getSeconds();
    }

    //
    // Return true if the sensor should be turned on. I.e., now falls
    // within the startSeconds / endSeconds range.
    //
    sensorOnTime(sensor) {
        let currentSeconds = this.currentSeconds();

        let turnOn = false;

        // Examine each of the defined ranges
        for (let i=0; i<sensor.activeSeconds.length && turnOn == false; i++) {
            let activeInterval = sensor.activeSeconds[i];

            this.log("[" + sensor.name + "]" + "[" + i + "] [" + this.secondsToHourMinSec(activeInterval.start) + " - " + this.secondsToHourMinSec(activeInterval.end) + "] --> " + this.secondsToHourMinSec(currentSeconds));

            // start / end w/in same day, e.g. 08:00 - 20:00
            if (activeInterval.start <= activeInterval.end) {
                turnOn = (activeInterval.start < currentSeconds) && (currentSeconds < activeInterval.end);
                sensor.activePeriod = i;
            }

            // start / end span days, e.g. 20:00 - 08:00
            else {
                turnOn = (activeInterval.start < currentSeconds) || (currentSeconds < activeInterval.end);
                sensor.activePeriod = i;
            }
        }

        if (!turnOn) {
            delete sensor.activePeriod;
        }

        return turnOn;
    }

    //
    // Turn sensor on after a random amount of off time.
    //
    startOnTimer(id) {
        const serviceMotion = this.serviceMotions[id];
        const serviceState = this.serviceStates[id];
        const sensor = this.sensors[id];

        let time = parseInt(Math.floor(Math.random() * (sensor.maxOffTime - sensor.minOffTime + 1) + sensor.minOffTime));
        let currentSeconds = this.currentSeconds();
        this.log("Starting on timer for sensor: " + id + ", delay: " + time +
                 " [" + this.secondsToHourMinSec(currentSeconds + time) + "]");

        serviceState.timeout = setTimeout(function() {
            // Only turn on sensors during allowed times
            if (this.sensorOnTime(sensor)) {
                this.log("Turning motion on for sensor: " + id);
                this.setSensorOn(id, true);
            }

            this.startOffTimer(id);
        }.bind(this), time*this.multiplier);
    }

    //
    // Time (in seconds) until end of active interval. 0 if not in
    // an active interval.
    //
    timeUntilEndOfInterval(sensor) {
        let currentSeconds = this.currentSeconds();
        let time = 0;

        let activeInterval = sensor.activeSeconds[sensor.activePeriod];

        // start / end w/in same day, e.g. 08:00 - 20:00
        if (activeInterval.start <= activeInterval.end) {
            if ((activeInterval.start < currentSeconds) && (currentSeconds < activeInterval.end)) {
                time = activeInterval.end - currentSeconds;
            }
        }

        // start / end span days, e.g. 20:00 - 08:00
        else {
            if ((activeInterval.start < currentSeconds) || (currentSeconds < activeInterval.end)) {
                const secondsPerDay = 86400;

                // before midnight --> seconds remaining in current day + seconds in next day
                if (activeInterval.start < currentSeconds) {
                    time = (secondsPerDay - currentSeconds) + activeInterval.end;
                }
                
                // after midnight --> seconds remaining in current day
                else {
                    time = activeInterval.end - currentSeconds;
                }
            }
        }

        return time;
    }

    //
    // Compute the off time at which to turn the given sensor off.
    //
    computeOffTime(sensor) {
        let time = parseInt(Math.floor(Math.random() * (sensor.maxOnTime - sensor.minOnTime + 1) + sensor.minOnTime));

        // check for a hard off time and adjust as needed
        if ((typeof sensor.activePeriod !== 'undefined') && sensor.activeTimesForSensor[sensor.activePeriod].absolute) {
            let maxTime = this.timeUntilEndOfInterval(sensor);
            if (time > maxTime) {
                time = maxTime;
            }
        }

        return time;
    }

    //
    // Turn sensor off after a random amount of on time.
    //
    startOffTimer(id) {
        const serviceMotion = this.serviceMotions[id];
        const serviceState = this.serviceStates[id];
        const sensor = this.sensors[id];

        let time = this.computeOffTime(sensor);
        let currentSeconds = this.currentSeconds();
        this.log("Starting off timer for sensor: " + id + ", delay: " + time +
                 " [" + this.secondsToHourMinSec(currentSeconds + time) + "]");

        serviceState.timeout = setTimeout(function() {
            this.log("Turning motion off for sensor: " + id);
            this.setSensorOn(id, false);

            this.startOnTimer(id);
        }.bind(this), time*this.multiplier);
    }

    //
    // Initialize the sensor.
    //
    async startSensor(id) {
        this.log("Starting sensor: " + id);

        // restore flag set for this sensor - check it - start up only
        if (this.sensors[id].restore) {
            let sensor = this.sensors[id];
            let isOn = await storage.getItem(sensor.name);

            // we only do this once
            delete sensor.restore;

            // On when we quit and still can be on
            if (isOn === 'true' && this.sensorOnTime(sensor)) {
                this.log(`Restore sensor [${sensor.name}] to on`);

                // restore sensor to motion detected
                this.log("Turning motion on for sensor: " + id);
                this.setSensorOn(id, true);

                // start 'off' timer
                this.startOffTimer(id);
            }

            // On when we quit, but shouldn't be on now
            // Note: We trigger it on first, then turn it off,
            //       otherwise the off won't be sent out because
            //       the homebridge/hap (?) framework thinks it's off.
            else if (isOn === 'true') {
                this.log("Turning motion off for sensor: " + id);

                // turn it on first
                this.setSensorOn(id, true);

                // then turn it off after delay and start on timer
                setTimeout(function(idx, on) {
                    this.setSensorOn(idx, on);
                    this.startOnTimer(idx);
                }.bind(this), 15000, id, false);
            }

            // Off when we quit - delay before turning on
            else {
                this.startOnTimer(id);
            }
        }
        
        // standard start for sensor - delay before turning on
        else {
            this.startOnTimer(id);
        }
    }

    //
    // Shut down the sensor. Clear timeouts. Turn if off.
    //
    stopSensor(id) {
        this.log("Stopping sensor: " + id);

        const serviceState = this.serviceStates[id];
        const motionDetected = serviceState.motionDetected;
        const timeout = serviceState.timeout;

        // Timeout currently set, cancel
        if (timeout) {
            this.log("Stopping timer for sensor: " + id);
            clearTimeout(timeout);
            delete serviceState.timeout;
        }

        // Sensor is currently on, turn it off
        if (motionDetected) {
            this.log("Turning motion off for sensor: " + id);
            this.setSensorOn(id, false);
        }
    }

    //
    // Activate or deactivate a sensor.
    // 
    async setSensorOn(id, on) {
        this.serviceMotions[id].updateCharacteristic(Characteristic.MotionDetected, on);
        this.serviceStates[id].motionDetected = on;
        await storage.setItem(this.sensors[id].name, on ? 'true' : 'false');
    }

    //
    // Turn switch on or off.
    //
    async setOn(on, callback) {

        // Store the new state
        await storage.setItem('active', on ? 'true' : 'false');

        // Turn away mode on, currently off
        if (on && !this.isSwitchOn) {
            this.log("Switch turned on");

            //Turn on the switch
            this.isSwitchOn = true;

            // Turn on the sensors
            for (let x = 0; x < this.sensors.length; x++) {
                this.startSensor(x);
            }
        }

        // Turn away mode off, currently on
        else if (!on && this.isSwitchOn) {
            this.log("Switch turned off");

            // Turn off the switch
            this.isSwitchOn = false;

            // Turn off the sensors
            for (let x = 0; x < this.sensors.length; x++) {
                this.stopSensor(x);
            }
        }

        callback && callback();
    }

    //
    // Share services that have been created.
    //
    getServices() {
        return [this.serviceSwitch, ...this.serviceMotions];
    }
}
